### ISSUE TITLE
Project.AssemblyName does not accurately report assembly name.

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#42465]
_This issue has a corresponding [ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/943329/projectassemblyname-does-not-accurately-report-ass.html). Please vote and comment there to make sure your voice is heard._

---
The following code suddently started appending ".csproj" onto assembly names which seems wrong.

```c#
private IReadOnlyList<string> GetTargetAssemblies()
{
    List<string> result = new List<string>();

    foreach (Microsoft.CodeAnalysis.Project project in this.workspace.CurrentSolution.Projects)
    {
        // .Net Core projects can have ".csproj" attached sometimes.
        result.Add(project.AssemblyName.Replace(".csproj", string.Empty));
    }

    return result.AsReadOnly();
}
```

It is important to keep apis reporting data in a consistent format. if it doesn't it requires  rolling out extension updates for no useful reason.

Here is the csproj that causes the assembly name to be wrong:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <AssemblyName>Microsoft.Exchange.Diagnostics</AssemblyName>
  </PropertyGroup>
</Project>
```

Note: I must have `<AssemblyName>` in this new format project because there is no other way (known? supported?) to find the target assembly programmatically between old and new format csproj files.

**Comment by sharwell** [comment]
@nschuessler I believe you are working on this?

**Comment by nschuessler** [comment]
> @nschuessler I believe you are working on this?

I'm gonna give it a try, yes.

**Comment by nschuessler** [comment]
Ok. With the help of @jasonmalinowski we know the main problem. The actual repro requires the build file to be like this:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <AssemblyName>Microsoft.Exchange.Diagnostics</AssemblyName>
  </PropertyGroup>

  <ItemGroup>
    <Compile Include="Exwatson.cs" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="System.Collections.Immutable" Version="1.7.0" />
  </ItemGroup>
</Project>
```
With the file on the windows file system with a different casing as 'ExWatson.cs`.

msbuild passes two files with different casings (i.e. unix file system casing rules) to csc.exe, which says they are the same file (which is true) and fails the build.

It seems the `msbuild` and `csc.exe` should agree on an algorithm to determine if files are the same or not.

**Comment by nschuessler** [comment]
Here is the output of csc.exe btw:
```C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Roslyn\Microsoft.CSharp.Core.targets(59,5): error MSB3105: The item "Exwatson.cs" was specified more than once in the "Sources" parameter.  Duplicate items are not supported by the "Sources" parameter.
```

**Comment by jasonmalinowski** [comment]
Actually looking at that error now, is MSBuild just blocking it? @rainersigwald or @jaredpar any idea what is actually emitting that error from @nschuessler's comment immediately above?

**Comment by rainersigwald** [comment]
It looks like the error is being emitted here:

https://github.com/dotnet/roslyn/blob/79770acfe36c099772465093389b013fc83681df/src/Compilers/Core/MSBuildTask/ManagedCompiler.cs#L1008

The string definition is copy/pasted from an MSBuild string, but AFAICT it's never used in the MSBuild repo.

The root problem here is that default items are enabled (by default), so a `**/*.cs` glob is being expanded to the correct case on disk, and then the explicit include with different case is pulled in. If an SDK project wants to manage source files explicitly, it must set

```xml
<EnableDefaultCompileItems>false</EnableDefaultCompileItems>
```

**Comment by nschuessler** [comment]
Q: Why does Roslyn(?) or project system allow the build to proceed so far that it gets into the case where we hit this problem.  IMHO, if this is strict requirement:
1) We should automatically set `EnableDefaultCompileItems = false` if they see `<Compile Include="xxx" />. This makes usability much easier for customer (i.e. we don't have internet search for secret feature on/off switches)
2) We should have file system sensitivity based on operating system (e.g. OrdinalIgnoreCase or Ordinal) so the system works.

**Comment by jaredpar** [comment]
@nschuessler 

>  Why does Roslyn(?) or project system allow the build to proceed so far that it gets into the case where we hit this problem

This is effectively the earliest the error can be detected. The C# compiler does not allow duplicate files to be provided as command line arguments. The task associated with the compiler is detecting and reporting this before invoking the compiler. There is no other place earlier in the pipeline where this could be accurately reported. 

MSBuild can't enforce this earlier because it's a restriction imposed by the C# compiler, not MSBuild. There are likely many tasks out there which allow duplicate items to be processed. Hence MSBuild must let this happen. 

Where do you propose this checking should be done eralier?

**Comment by nschuessler** [comment]
I was thinking when the csproj is loaded / parsed it should detect it and report back to the caller. In fact it does, it is just a warning/info error methinks.  Is msbuild what is calling Roslyn to do the compile at that point?

**Comment by jaredpar** [comment]
> I was thinking when the csproj is loaded / parsed it should detect it and report back to the caller. 

When the project is parsed it's unknown if the end task being called is the C# compiler. Or if there is a task earlier in the pipeline which removes duplicates. Or if there is a task earlier which deletes the C# compiler task. All of these are possibilitie. MSBuild is infinitely extensible. 

The earliest point we can know if the C# compiler is going to actually be invoked is when we're in the C# compiler task. That is when the error is reported.

**Comment by rainersigwald** [comment]
> IMHO, if this is strict requirement:
> 
> 1. We should automatically set `EnableDefaultCompileItems = false` if they see
> 
> ```
> <Compile Include="xxx" />
> ```

This isn't the requirement. It's perfectly ok to have

```xml
<Compile Include="..\..\shared\SomeSharedSourceFile.cs" />
```

in a project file, even with `EnableDefaultCompileItems` on (the default), because the explicit include there doesn't overlap with the default glob (`**\*.cs`, starting from the directory the `.csproj` is in). There are also other includes which wouldn't overlap and are thus fine.

> We should have file system sensitivity based on operating system

Alas, it's not that easy. To be completely accurate, it'd have to be based on _file_ system, which can vary in complex ways--across directories, across OS versions, across a single OS version based on setting.

**Comment by jaredpar** [comment]
> Alas, it's not that easy. To be completely accurate, it'd have to be based on file system, which can vary in complex ways--across directories, across OS versions, across a single OS version based on setting

Indeed. It's possible to have directory paths where different folders have different case sensitivity requirements. This oddness is not limited to a particular operating system either. Basically paths have no predictable case sensitivity requirements.

**Comment by nschuessler** [comment]
So can a build task be inserted before the C# task that tests the environment where the specified files are to see if there is case sensitivity or not?  Then de-duplicate if needed? There are other build systems that discover a lot more complicated stuff (kernel configuration on Linux) before the build than this.

**Comment by nschuessler** [comment]
Maybe this is the solution.
1) `AssemblyName` is not reliable (useful) property on Project and should be removed / deprecated. By having the property there it implies it can be used for something.
2) Since `AssemblyName` can't reliability determined, the diagnostic tool/pipelines should not use `AssemblyName` to uniquely identify symbol names.  I.e. mscorlib!System.Object should be <guid>!System.Object, where GUID is specified in AssemblyInfo.cs.

As an alternative, maybe embedded somewhere (Assembly manifest, symbol files) should be the MD5 hash of the method body, and MD5 hash of the method signature that can be extracted from dump files and ETW traces of GC.

The problem with symbols files is that it is quite difficult to maintain matching symbol files with binaries in the data center... Even when you package .NET binaries into a nuget package …  So it would be preferable if the hashes were in the metadata inside the assembly

**Comment by nschuessler** [comment]
Actually, the project properties page has it populated correctly. Is there another API to retrieve this information?

![image](https://user-images.githubusercontent.com/54370830/77354763-94acc680-6d00-11ea-9a29-e728e7ca1f73.png)

**Comment by jmarolf** [comment]
> AssemblyName is not reliable (useful) property on Project and should be removed / deprecated. By having the property there it implies it can be used for something.

The issue is that msbuild is so extensible that anyone can change that value. The project property page is calling the same apis as everyone else (its also open sourced [here](https://github.com/dotnet/project-system)). You should not depend on that property being in any particular shape since it has not been throughout the history of .NET

**Comment by jasonmalinowski** [comment]
> @jaredpar: The earliest point we can know if the C# compiler is going to actually be invoked is when we're in the C# compiler task. That is when the error is reported.

So in this case the task would have been told SkipCompilerExecution = true, so the compiler wouldn't have gotten invoked in the design time build. Is it possible to skip the checks if that's set and the other flag to provide the command line string is still active? That way real builds aren't impacted but the project system can still get what it needs here.

> Actually, the project properties page has it populated correctly. Is there another API to retrieve this information?

You could try, but as @jmarolf observes there's no promise that value is correct in general. And given the project you are looking at _doesn't even build_, I'm not sure why it's worth trying to handle that error case. As your screenshot shows: there's a gold bar saying the solution didn't load. At that point, the warranty is void...

**Comment by jasonmalinowski** [comment]
> That way real builds aren't impacted but the project system can still get what it needs here.

And oh: that would apply to any sort of argument checking being done in this way if feasable. Basically, if SkipCompilerExecution is on, you're not actually doing a build so some fuzziness might be OK.

**Comment by nschuessler** [comment]
@jasonmalinowski  The gold bar is for another project (xxx.common), not the one in question that doesn't load.  The project that gives no errors until the build and an incorrect assembly name is "Microsoft.Exchange.Diagnostics".  

But the IDE does need to work even if things doesn't build, and roslyn works that way (if I recall). If you have a syntax error then it gives you as much info as it can so IDE features can work.

```
public void DoSomething()
{
      while (!done)  // Not declared
      {
           x == y  // Syntax error
      }
}
```

Just because there is a syntax error, should roslyn return nothing from MethodDeclarationSyntax?  I believe it does so that things like "Rename" refactoring still work. Otherwise the UI would just lock everything down and behave like notepad.

**Comment by jaredpar** [comment]
@jasonmalinowski 

> Is it possible to skip the checks if that's set and the other flag to provide the command line string is still active? 

Sure that's possible. But then I'd push back why do we raise **any** error during our target evaluation if it's a design time build?  Doesn't practically every error we can produce in our task before going to the compiler meeting this general idea? If so I have to wonder why for 20 years it's been done this way.

I really feel like we're getting a bit far afield here though. There is an error in the project file, the design time build process correctly identified the error and it's correctable by the user.

**Comment by nschuessler** [comment]
@jaredpar  It identified the error, but as a user, I had to get some special logs on the interaction between various undocumented interactions between build pipeline components to get the error.  

As in response to @jasonmalinowski  (above) it is strange that I can get namespace/class/method information reliably but not the enclosing assembly target.

**Comment by nschuessler** [comment]
@jaredpar  So if I create a build task that runs after the C# task (is it [Csc](https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.tasks.csc.-ctor?view=netframework-4.8)?), can I get the target name after it succeeds building?

What properties on the CSC task do I need to look at to get the csproj, and the target binary name?

Basically, if I can get `.csproj` and `targetname` I can create a giant CSV with:
build#,repo path to .csproj,TargetBinaryName

and put it in an azure table to query to map csproj -> targetfilename.

**Comment by jaredpar** [comment]
> build#,repo path to .csproj,TargetBinaryName

If you have multi-targeting enabled then a single project file will build multiple binaries. They should all have the same name, unless you did some customizations, but will have different relative output paths. 

> What properties on the CSC task do I need to look at to get the csproj, and the target binary name?

The most reliable way to determine the binary name is to look at the output of the target `CoreCompile` which is named `IntermediateAssembly`. That will contain the file path of the assembly the compiler wrote to disk. 

Note though: this is not considered a public API and is subject to change at any time.

**Comment by jasonmalinowski** [comment]
> Doesn't practically every error we can produce in our task before going to the compiler meeting this general idea? If so I have to wonder why for 20 years it's been done this way.

Looking further it looks like the Sources and Resources item groups are special:

https://github.com/dotnet/roslyn/blob/79770acfe36c099772465093389b013fc83681df/src/Compilers/Core/MSBuildTask/ManagedCompiler.cs#L960

Those are validated *there* which means the rest of the process is immediately stopped. The rest of the errors are happening after the command line has been generated and thus the process might have continued fine. Also note that we only recently moved to command line outputting: for the first 19 years everything was going through the host object that might have been pushed to even if other errors are logged. (At least looking at our _current_ code just that duplicate checking would have happened earlier.)

People widely using globbing is also newer, of course.

**Comment by jaredpar** [comment]
@jasonmalinowski 

> Looking further it looks like the Sources and Resources item groups are special

One other aspect in which they're special: the enforcement of no duplicate items is also done by the compiler. 

https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Compilation/Compilation.cs#L1688

This means we're doing duplicate checking here and I can't discern why that is. Generally if it's important for compliation success then we do the check in `csc`. After all the compiler can be invoked directly hence we have to assume we do all the error checking. 

@rainersigwald can you think of any particular reason why such parameter validation should be duplicated in the MSBuild layer? My instinct here is to just delete this code and rely on the compilec checking to do the right thing. 

> Those are validated there which means the rest of the process is immediately stopped

That is odd. I'm unsure what makes these errors special. From a compiler perspective it's just another piece of command line validation. Is it possible there is something about design time builds that require this to be done?

> Also note that we only recently moved to command line outputting: for the first 19 years everything was going through the host object that might have been pushed to even if other errors are logged.

Don't think this is related. Took a peak through the `ToolTask` source and the host object usage is also predicated on `ValidateParameters` succeeding.

**Comment by rainersigwald** [comment]
> @rainersigwald can you think of any particular reason why such parameter validation should be duplicated in the MSBuild layer? My instinct here is to just delete this code and rely on the compilec checking to do the right thing.

Nothing springs to mind. That sounds like a fine plan.
